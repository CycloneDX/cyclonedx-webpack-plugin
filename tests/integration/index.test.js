/*!
This file is part of CycloneDX Webpack plugin.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

SPDX-License-Identifier: Apache-2.0
Copyright (c) OWASP Foundation. All Rights Reserved.
*/

const fs = require('fs')
const path = require('path')
const { spawnSync } = require('child_process')

const { describe, expect, it } = require('@jest/globals')

const { version: thisVersion } = require('../../package.json')

const testSetups = [
  // region functional
  {
    dir: 'webpack5-vue2',
    purpose: 'functional: webpack5 with vue2',
    results: [ // paths relative to `dir`
      {
        format: 'xml',
        file: 'dist/.bom/bom.xml'
      },
      {
        format: 'json',
        file: 'dist/.bom/bom.json'
      },
      {
        format: 'json',
        file: 'dist/.well-known/sbom'
      }
    ]
  },
  {
    dir: 'webpack5-angular13',
    purpose: 'functional: webpack5 with angular13',
    results: [ // paths relative to `dir`
      {
        format: 'xml',
        file: 'dist/.bom/bom.xml'
      },
      {
        format: 'json',
        file: 'dist/.bom/bom.json'
      },
      {
        format: 'json',
        file: 'dist/.well-known/sbom'
      }
    ]
  },
  {
    dir: 'webpack5-angular17',
    purpose: 'functional: webpack5 with angular17',
    results: [ // paths relative to `dir`
      {
        format: 'xml',
        file: 'dist/.bom/bom.xml'
      },
      {
        format: 'json',
        file: 'dist/.bom/bom.json'
      },
      {
        format: 'json',
        file: 'dist/.well-known/sbom'
      }
    ]
  },
  {
    dir: 'webpack5-react18',
    purpose: 'functional: webpack5 with react18',
    results: [ // paths relative to `dir`
      {
        format: 'xml',
        file: 'dist/.bom/bom.xml'
      },
      {
        format: 'json',
        file: 'dist/.bom/bom.json'
      },
      {
        format: 'json',
        file: 'dist/.well-known/sbom'
      }
    ]
  },
  // endregion functional
  // region regression
  {
    dir: 'regression-issue745',
    purpose: 'regression: issue#745',
    results: [ // paths relative to `dir`
      {
        format: 'xml',
        file: 'dist/.bom/bom.xml'
      },
      {
        format: 'json',
        file: 'dist/.bom/bom.json'
      },
      {
        format: 'json',
        file: 'dist/.well-known/sbom'
      }
    ]
  }
  // endregion regression
]

// for testing purposes, some outdated jest version must be used.
// this version has a different format for snapshots ...
let compareSnapshots
try {
  compareSnapshots = Number(require('jest/package.json').version.split('.')[0]) >= 29
} catch {
  compareSnapshots = null
}

describe('integration', () => {
  testSetups.forEach(({ purpose, dir, results }) => {
    const setupMethod = 'yarn'  // @TODO detect yarn / pnpm / ... fallback npm

    describe(purpose, () => {
      const built = spawnSync(
        setupMethod, ['run', 'build'], {
          cwd: path.resolve(module.path, dir),
          stdio: ['ignore', 'pipe', 'pipe'],
          encoding: 'utf8',
          shell: true,
          env: {
            PATH: process.env.PATH,
            CI: '1'
          }
        }
      )
      let skipTests = false
      try {
        expect(built.status).toBe(0)
      } catch (err) {
        if (/should not be used for production|Angular CLI requires a minimum|does not support Node\.js v/.test(built.stderr.toString())) {
          skipTests = true
        } else {
          console.log(built, '\n')
          throw err
        }
      }

      results.forEach(({ format, file }) => {
        (skipTests
          ? it.skip
          : it
        )(`generated ${format} file: ${file}`, () => {
          const resultFile = path.resolve(module.path, dir, file)
          const resultBuffer = fs.readFileSync(resultFile)
          expect(resultBuffer).toBeInstanceOf(Buffer)
          expect(resultBuffer.length).toBeGreaterThan(0)
          const resultReproducible = makeReproducible(format, resultBuffer.toString())
          if (compareSnapshots) {
            expect(resultReproducible).toMatchSnapshot()
          }
        })
      })
    })
  })
})

/**
 * @param {string} format
 * @param {*} data
 * @returns {string}
 */
function makeReproducible (format, data) {
  switch (format.toLowerCase()) {
    case 'xml':
      return makeXmlReproducible(data)
    case 'json':
      return makeJsonReproducible(data)
    default:
      throw new RangeError(`unexpected format: ${format}`)
  }
}

/**
 * @param {string} json
 * @returns {string}
 */
function makeJsonReproducible (json) {
  return json
    .replace(
      // replace metadata.tools.version
      '        "vendor": "@cyclonedx",\n' +
      '        "name": "webpack-plugin",\n' +
      `        "version": ${JSON.stringify(thisVersion)},\n`,
      '        "vendor": "@cyclonedx",\n' +
      '        "name": "webpack-plugin",\n' +
      '        "version": "thisVersion-testing",\n'
    ).replace(
      // replace metadata.tools.version
      new RegExp(
        '        "vendor": "@cyclonedx",\n' +
        '        "name": "cyclonedx-library",\n' +
        '        "version": ".+?",\n'
      ),
      '        "vendor": "@cyclonedx",\n' +
      '        "name": "cyclonedx-library",\n' +
      '        "version": "libVersion-testing",\n'
    )
}

/**
 * @param {string} xml
 * @returns {string}
 *
 * eslint-disable-next-line no-unused-vars
 */
function makeXmlReproducible (xml) {
  return xml
    .replace(
      // replace metadata.tools.version
      '        <vendor>@cyclonedx</vendor>\n' +
      '        <name>webpack-plugin</name>\n' +
      `        <version>${thisVersion}</version>`,
      '        <vendor>@cyclonedx</vendor>\n' +
      '        <name>webpack-plugin</name>\n' +
      '        <version>thisVersion-testing</version>'
    ).replace(
      // replace metadata.tools.version
      new RegExp(
        '        <vendor>@cyclonedx</vendor>\n' +
        '        <name>cyclonedx-library</name>\n' +
        '        <version>.+?</version>'
      ),
      '        <vendor>@cyclonedx</vendor>\n' +
      '        <name>cyclonedx-library</name>\n' +
      '        <version>libVersion-testing</version>'
    )
}
