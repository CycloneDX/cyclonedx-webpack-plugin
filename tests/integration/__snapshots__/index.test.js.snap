// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration functional: webpack5 with vue2 in pnpm setup generated json file: dist/.bom/bom.json 1`] = `
"{
  "$schema": "http://cyclonedx.org/schema/bom-1.6.schema.json",
  "bomFormat": "CycloneDX",
  "specVersion": "1.6",
  "version": 1,
  "metadata": {
    "lifecycles": [
      {
        "phase": "build"
      }
    ],
    "tools": [
      {
        "vendor": "@cyclonedx",
        "name": "cyclonedx-library",
        "version": "libVersion-testing",
        "externalReferences": [
          {
            "url": "https://github.com/CycloneDX/cyclonedx-javascript-library/issues",
            "type": "issue-tracker",
            "comment": "as detected from PackageJson property \\"bugs.url\\""
          },
          {
            "url": "git+https://github.com/CycloneDX/cyclonedx-javascript-library.git",
            "type": "vcs",
            "comment": "as detected from PackageJson property \\"repository.url\\""
          },
          {
            "url": "https://github.com/CycloneDX/cyclonedx-javascript-library#readme",
            "type": "website",
            "comment": "as detected from PackageJson property \\"homepage\\""
          }
        ]
      },
      {
        "vendor": "@cyclonedx",
        "name": "webpack-plugin",
        "version": "thisVersion-testing",
        "externalReferences": [
          {
            "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin/issues",
            "type": "issue-tracker",
            "comment": "as detected from PackageJson property \\"bugs.url\\""
          },
          {
            "url": "git+https://github.com/CycloneDX/cyclonedx-webpack-plugin.git",
            "type": "vcs",
            "comment": "as detected from PackageJson property \\"repository.url\\""
          },
          {
            "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin#readme",
            "type": "website",
            "comment": "as detected from PackageJson property \\"homepage\\""
          }
        ]
      }
    ],
    "component": {
      "type": "application",
      "name": "example-webpack5-vue2-pnpm",
      "group": "@cyclonedx-webpack-plugin-tests",
      "bom-ref": "pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
      "author": "Jan Kowalleck",
      "description": "example setup witch Vue2 in WebPack5 with pnpm setup",
      "licenses": [
        {
          "license": {
            "id": "Apache-2.0",
            "acknowledgement": "declared"
          }
        }
      ],
      "purl": "pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
      "externalReferences": [
        {
          "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin/issues",
          "type": "issue-tracker",
          "comment": "as detected from PackageJson property \\"bugs.url\\""
        },
        {
          "url": "git+https://github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
          "type": "vcs",
          "comment": "as detected from PackageJson property \\"repository.url\\" and \\"repository.directory\\""
        },
        {
          "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin/tree/master/tests/integration/webpack5-vue2#readme",
          "type": "website",
          "comment": "as detected from PackageJson property \\"homepage\\""
        }
      ]
    }
  },
  "components": [
    {
      "type": "library",
      "name": "vue",
      "version": "2.6.14",
      "bom-ref": "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git",
      "author": "Evan You",
      "description": "Reactive, component-oriented view layer for modern web interfaces.",
      "licenses": [
        {
          "license": {
            "id": "MIT",
            "acknowledgement": "declared"
          }
        }
      ],
      "purl": "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git",
      "externalReferences": [
        {
          "url": "https://github.com/vuejs/vue/issues",
          "type": "issue-tracker",
          "comment": "as detected from PackageJson property \\"bugs.url\\""
        },
        {
          "url": "git+https://github.com/vuejs/vue.git",
          "type": "vcs",
          "comment": "as detected from PackageJson property \\"repository.url\\""
        },
        {
          "url": "https://github.com/vuejs/vue#readme",
          "type": "website",
          "comment": "as detected from PackageJson property \\"homepage\\""
        }
      ]
    }
  ],
  "dependencies": [
    {
      "ref": "pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
      "dependsOn": [
        "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git"
      ]
    },
    {
      "ref": "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git"
    }
  ]
}"
`;

exports[`integration functional: webpack5 with vue2 in pnpm setup generated json file: dist/.well-known/sbom 1`] = `
"{
  "$schema": "http://cyclonedx.org/schema/bom-1.6.schema.json",
  "bomFormat": "CycloneDX",
  "specVersion": "1.6",
  "version": 1,
  "metadata": {
    "lifecycles": [
      {
        "phase": "build"
      }
    ],
    "tools": [
      {
        "vendor": "@cyclonedx",
        "name": "cyclonedx-library",
        "version": "libVersion-testing",
        "externalReferences": [
          {
            "url": "https://github.com/CycloneDX/cyclonedx-javascript-library/issues",
            "type": "issue-tracker",
            "comment": "as detected from PackageJson property \\"bugs.url\\""
          },
          {
            "url": "git+https://github.com/CycloneDX/cyclonedx-javascript-library.git",
            "type": "vcs",
            "comment": "as detected from PackageJson property \\"repository.url\\""
          },
          {
            "url": "https://github.com/CycloneDX/cyclonedx-javascript-library#readme",
            "type": "website",
            "comment": "as detected from PackageJson property \\"homepage\\""
          }
        ]
      },
      {
        "vendor": "@cyclonedx",
        "name": "webpack-plugin",
        "version": "thisVersion-testing",
        "externalReferences": [
          {
            "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin/issues",
            "type": "issue-tracker",
            "comment": "as detected from PackageJson property \\"bugs.url\\""
          },
          {
            "url": "git+https://github.com/CycloneDX/cyclonedx-webpack-plugin.git",
            "type": "vcs",
            "comment": "as detected from PackageJson property \\"repository.url\\""
          },
          {
            "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin#readme",
            "type": "website",
            "comment": "as detected from PackageJson property \\"homepage\\""
          }
        ]
      }
    ],
    "component": {
      "type": "application",
      "name": "example-webpack5-vue2-pnpm",
      "group": "@cyclonedx-webpack-plugin-tests",
      "bom-ref": "pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
      "author": "Jan Kowalleck",
      "description": "example setup witch Vue2 in WebPack5 with pnpm setup",
      "licenses": [
        {
          "license": {
            "id": "Apache-2.0",
            "acknowledgement": "declared"
          }
        }
      ],
      "purl": "pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
      "externalReferences": [
        {
          "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin/issues",
          "type": "issue-tracker",
          "comment": "as detected from PackageJson property \\"bugs.url\\""
        },
        {
          "url": "git+https://github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
          "type": "vcs",
          "comment": "as detected from PackageJson property \\"repository.url\\" and \\"repository.directory\\""
        },
        {
          "url": "https://github.com/CycloneDX/cyclonedx-webpack-plugin/tree/master/tests/integration/webpack5-vue2#readme",
          "type": "website",
          "comment": "as detected from PackageJson property \\"homepage\\""
        }
      ]
    }
  },
  "components": [
    {
      "type": "library",
      "name": "vue",
      "version": "2.6.14",
      "bom-ref": "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git",
      "author": "Evan You",
      "description": "Reactive, component-oriented view layer for modern web interfaces.",
      "licenses": [
        {
          "license": {
            "id": "MIT",
            "acknowledgement": "declared"
          }
        }
      ],
      "purl": "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git",
      "externalReferences": [
        {
          "url": "https://github.com/vuejs/vue/issues",
          "type": "issue-tracker",
          "comment": "as detected from PackageJson property \\"bugs.url\\""
        },
        {
          "url": "git+https://github.com/vuejs/vue.git",
          "type": "vcs",
          "comment": "as detected from PackageJson property \\"repository.url\\""
        },
        {
          "url": "https://github.com/vuejs/vue#readme",
          "type": "website",
          "comment": "as detected from PackageJson property \\"homepage\\""
        }
      ]
    }
  ],
  "dependencies": [
    {
      "ref": "pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn",
      "dependsOn": [
        "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git"
      ]
    },
    {
      "ref": "pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git"
    }
  ]
}"
`;

exports[`integration functional: webpack5 with vue2 in pnpm setup generated xml file: dist/.bom/bom.xml 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<bom xmlns="http://cyclonedx.org/schema/bom/1.6" version="1">
  <metadata>
    <lifecycles>
      <lifecycle>
        <phase>build</phase>
      </lifecycle>
    </lifecycles>
    <tools>
      <tool>
        <vendor>@cyclonedx</vendor>
        <name>cyclonedx-library</name>
        <version>libVersion-testing</version>
        <externalReferences>
          <reference type="issue-tracker">
            <url>https://github.com/CycloneDX/cyclonedx-javascript-library/issues</url>
            <comment>as detected from PackageJson property "bugs.url"</comment>
          </reference>
          <reference type="vcs">
            <url>git+https://github.com/CycloneDX/cyclonedx-javascript-library.git</url>
            <comment>as detected from PackageJson property "repository.url"</comment>
          </reference>
          <reference type="website">
            <url>https://github.com/CycloneDX/cyclonedx-javascript-library#readme</url>
            <comment>as detected from PackageJson property "homepage"</comment>
          </reference>
        </externalReferences>
      </tool>
      <tool>
        <vendor>@cyclonedx</vendor>
        <name>webpack-plugin</name>
        <version>thisVersion-testing</version>
        <externalReferences>
          <reference type="issue-tracker">
            <url>https://github.com/CycloneDX/cyclonedx-webpack-plugin/issues</url>
            <comment>as detected from PackageJson property "bugs.url"</comment>
          </reference>
          <reference type="vcs">
            <url>git+https://github.com/CycloneDX/cyclonedx-webpack-plugin.git</url>
            <comment>as detected from PackageJson property "repository.url"</comment>
          </reference>
          <reference type="website">
            <url>https://github.com/CycloneDX/cyclonedx-webpack-plugin#readme</url>
            <comment>as detected from PackageJson property "homepage"</comment>
          </reference>
        </externalReferences>
      </tool>
    </tools>
    <component type="application" bom-ref="pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn">
      <author>Jan Kowalleck</author>
      <group>@cyclonedx-webpack-plugin-tests</group>
      <name>example-webpack5-vue2-pnpm</name>
      <description>example setup witch Vue2 in WebPack5 with pnpm setup</description>
      <licenses>
        <license acknowledgement="declared">
          <id>Apache-2.0</id>
        </license>
      </licenses>
      <purl>pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn</purl>
      <externalReferences>
        <reference type="issue-tracker">
          <url>https://github.com/CycloneDX/cyclonedx-webpack-plugin/issues</url>
          <comment>as detected from PackageJson property "bugs.url"</comment>
        </reference>
        <reference type="vcs">
          <url>git+https://github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn</url>
          <comment>as detected from PackageJson property "repository.url" and "repository.directory"</comment>
        </reference>
        <reference type="website">
          <url>https://github.com/CycloneDX/cyclonedx-webpack-plugin/tree/master/tests/integration/webpack5-vue2#readme</url>
          <comment>as detected from PackageJson property "homepage"</comment>
        </reference>
      </externalReferences>
    </component>
  </metadata>
  <components>
    <component type="library" bom-ref="pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git">
      <author>Evan You</author>
      <name>vue</name>
      <version>2.6.14</version>
      <description>Reactive, component-oriented view layer for modern web interfaces.</description>
      <licenses>
        <license acknowledgement="declared">
          <id>MIT</id>
        </license>
      </licenses>
      <purl>pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git</purl>
      <externalReferences>
        <reference type="issue-tracker">
          <url>https://github.com/vuejs/vue/issues</url>
          <comment>as detected from PackageJson property "bugs.url"</comment>
        </reference>
        <reference type="vcs">
          <url>git+https://github.com/vuejs/vue.git</url>
          <comment>as detected from PackageJson property "repository.url"</comment>
        </reference>
        <reference type="website">
          <url>https://github.com/vuejs/vue#readme</url>
          <comment>as detected from PackageJson property "homepage"</comment>
        </reference>
      </externalReferences>
    </component>
  </components>
  <dependencies>
    <dependency ref="pkg:npm/%40cyclonedx-webpack-plugin-tests/example-webpack5-vue2-pnpm?vcs_url=git%2Bhttps%3A//github.com/CycloneDX/cyclonedx-webpack-plugin.git#tests/integration/webpack5-vue2-yarn">
      <dependency ref="pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git"/>
    </dependency>
    <dependency ref="pkg:npm/vue@2.6.14?vcs_url=git%2Bhttps%3A//github.com/vuejs/vue.git"/>
  </dependencies>
</bom>"
`;
